source/game/field/CollisionDirector.cc

```diff
@@ -0,0 +1,82 @@
+#include "CollisionDirector.hh"
+
+#include "game/field/CourseColMgr.hh"
+
+namespace Field {
+
+CollisionDirector *CollisionDirector::CreateInstance() {
+    assert(!s_instance);
+    s_instance = new CollisionDirector;
+    return s_instance;
+}
+
+CollisionDirector *CollisionDirector::Instance() {
+    return s_instance;
+}
+
+void CollisionDirector::DestroyInstance() {
+    assert(s_instance);
+    delete s_instance;
+    s_instance = nullptr;
+}
+
+CollisionDirector::CollisionDirector() {
+    CourseColMgr::CreateInstance()->init();
+}
+
+CollisionDirector::~CollisionDirector() = default;
+
+bool CollisionDirector::checkSphereFullPush(f32 radius, const EGG::Vector3f &v0,
+        const EGG::Vector3f &v1, KCLTypeMask flags, CourseColMgr::CollisionInfo *pInfo,
+        KCLTypeMask *pFlagsOut, u32 /*param_8*/) {
+    if (pInfo) {
+        pInfo->m_bbox.mMin = EGG::Vector3f::zero;
+        pInfo->m_bbox.mMax = EGG::Vector3f::zero;
+        pInfo->_50 = -std::numeric_limits<f32>::min();
+        pInfo->m_wallDist = -std::numeric_limits<f32>::min();
+        pInfo->m_floorDist = -std::numeric_limits<f32>::min();
+        pInfo->m_perpendicularity = 0.0f;
+    }
+
+    if (pFlagsOut) {
+        resetCollisionEntries(pFlagsOut);
+    }
+
+    bool colliding = false;
+
+    if (flags &&
+            CourseColMgr::Instance()->checkSphereFullPush(1.0f, radius, nullptr, v0, v1, flags,
+                    pInfo, pFlagsOut)) {
+        colliding = true;
+    }
+
+    if (colliding && pInfo) {
+        pInfo->m_minPlusMax = pInfo->m_bbox.mMin + pInfo->m_bbox.mMax;
+    }
+
+    return colliding;
+}
+
+void CollisionDirector::resetCollisionEntries(u32 *ptr) {
+    *ptr = 0;
+    s_collisionEntryCount = 0;
+    s_closestCollisionEntry = 0;
+}
+
+void CollisionDirector::pushCollisionEntry(float dist, KCLTypeMask *typeMask,
+        KCLTypeMask kclTypeBit, u16 attribute) {
+    *typeMask = *typeMask | kclTypeBit;
+    if (s_collisionEntryCount >= s_entries.size()) {
+        s_collisionEntryCount = s_entries.size() - 1;
+    }
+
+    CollisionEntry entry(kclTypeBit, attribute, dist);
+    s_entries[s_collisionEntryCount++] = entry;
+}
+
+CollisionDirector *CollisionDirector::s_instance = nullptr;
+std::array<CollisionDirector::CollisionEntry, 0x40> CollisionDirector::s_entries = {};
+u32 CollisionDirector::s_collisionEntryCount = 0;
```

Comment:
We should use `size_t` to more accurately reflect its usecase.

Reply:
Alternatively we can reduce to `u8` given the array size, but I'm fine using `size_t`